### PR TITLE
aws-appmesh: Allow app ports to be optional

### PR DESCRIPTION
**Pull request by bcelenza**
*Description of changes:*

This is a cherry-pick of @kiranmeduri 's prior PR (#12) with incorporation of conflicting changes and minor updates on comment style.

*Testing:*

`make build e2e-test` only -- Do we have a documented way to test this in a real runtime environment?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/aws-appmesh/plugin/commands.go [hunk 1]
```diff
@@ -208,27 +206,70 @@ func (plugin *Plugin) setupIptablesRules(
 	return nil
 }
 
+// setupIngressRules installs iptable rules to handle ingress traffic.
+func (plugin *Plugin) setupIngressRules(iptable *iptables.IPTables,
+	config *config.NetConfig) error {
+	if config.ProxyIngressPort == "" {
```
Comment (ofiliz) [inline plugins/aws-appmesh/plugin/commands.go:212]: I understand you are checking for this misconfig case in netconfig validation, but do you think we should return if either one of ProxyIngressPort and AppPorts is empty here, just to be safe? If for whatever reason AppPorts is empty but ProxyIngressPort is not, line 222 will fail with a cryptic error.

File: plugins/aws-appmesh/plugin/commands.go [hunk 2]
```diff
@@ -208,27 +206,70 @@ func (plugin *Plugin) setupIptablesRules(
...
 	if err != nil {
-		log.Errorf("Delete the rule in OUTPUT chain failed: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// deleteIngressRules deletes the iptable rules for ingress traffic
+func (plugin *Plugin) deleteIngressRules(iptable *iptables.IPTables,
```
[the remark this answers is not in the document]
Comment (bcelenza) [reply plugins/aws-appmesh/plugin/commands.go]: Done.

File: plugins/aws-appmesh/plugin/commands.go [hunk 3]
```diff
@@ -208,27 +206,70 @@ func (plugin *Plugin) setupIptablesRules(
...
+	err = plugin.deleteEgressRules(iptable)
 	if err != nil {
-		log.Errorf("Delete the rule in OUTPUT chain failed: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// deleteIngressRules deletes the iptable rules for ingress traffic
```
[the remark this answers is not in the document]
Comment (bcelenza) [reply plugins/aws-appmesh/plugin/commands.go]: Done.

File: plugins/aws-appmesh/plugin/commands.go (hunk 1 above)
Comment (bcelenza) [reply plugins/aws-appmesh/plugin/commands.go]: Agree, added OR to catch empty AppPorts here.

File: plugins/aws-appmesh/config/netconfig.go [hunk 4]
```diff
@@ -129,16 +129,20 @@ func validateConfig(config netConfigJSON) error {
...
-		return fmt.Errorf("missing required parameter appPorts")
-	}
 	if config.ProxyEgressPort == "" {
 		return fmt.Errorf("missing required parameter proxyEgressPort")
 	}
-	if config.ProxyIngressPort == "" {
+
+	// AppPorts and ProxyIngressPort go in pairs,
+	// i.e. either both are set or both are not set.
+	if config.ProxyIngressPort == "" && len(config.AppPorts) > 0 {
```
Comment (yumex93) [inline plugins/aws-appmesh/config/netconfig.go:138]: nit: the error message is not clear enough for customers to know the relationship between appPorts and proxyIngressPort - either both set or not set.

File: plugins/aws-appmesh/plugin/commands.go [hunk 5]
```diff
@@ -208,27 +207,73 @@ func (plugin *Plugin) setupIptablesRules(
 	return nil
 }
 
+// setupIngressRules installs iptable rules to handle ingress traffic.
+func (plugin *Plugin) setupIngressRules(
+	iptable *iptables.IPTables,
+	config *config.NetConfig) error {
+	if config.ProxyIngressPort == "" || len(config.AppPorts) == 0 {
```
Comment (yumex93) [inline plugins/aws-appmesh/plugin/commands.go:214]: is **||** needed here? The only condition would happen is **&&**.

File: plugins/aws-appmesh/config/netconfig.go (hunk 4 above)
Comment (bcelenza) [reply plugins/aws-appmesh/config/netconfig.go:138]: Updated to hopefully be a little more clear. :)

File: plugins/aws-appmesh/plugin/commands.go (hunk 5 above)
Comment (bcelenza) [reply plugins/aws-appmesh/plugin/commands.go:214]: Admittedly this is a little over-protective, but if either one is not set, there will be an odd error message produced by this function (as @ofiliz noted above). In the event the function is used out-of-band, I felt an OR was better, as we only want to allow the ingress rules to be set up if *both* are set.

**Pull request merged**